<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>aggregateRegionCounts-methods — aggregateRegionCounts,DsATAC-method • ChrAccR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="aggregateRegionCounts-methods — aggregateRegionCounts,DsATAC-method"><meta property="og:description" content="Agregate counts across a set of regions, e.g. for footprinting analysis"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ChrAccR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/overview.html">Overview</a>
    </li>
    <li>
      <a href="../articles/singlecell.html">Single-cell analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/resource.html">Resource</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/EpigenomeInformatics/ChrAccR" class="external-link">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>aggregateRegionCounts-methods</h1>
    
    <div class="hidden name"><code>aggregateRegionCounts-DsATAC-method.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Agregate counts across a set of regions, e.g. for footprinting analysis</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for DsATAC</span></span>
<span><span class="fu">aggregateRegionCounts</span><span class="op">(</span></span>
<span>  <span class="va">.object</span>,</span>
<span>  <span class="va">regionGr</span>,</span>
<span>  samples <span class="op">=</span> <span class="fu"><a href="getSamples-DsAcc-method.html">getSamples</a></span><span class="op">(</span><span class="va">.object</span><span class="op">)</span>,</span>
<span>  countAggrFun <span class="op">=</span> <span class="st">"sum"</span>,</span>
<span>  norm <span class="op">=</span> <span class="st">"tailMean"</span>,</span>
<span>  normTailW <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  kmerBiasAdj <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  sampleCovg <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampleKmerFreqM <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  regionKmerFreqM <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  silent <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>.object</dt>
<dd><p><code><a href="DsATAC-class.html">DsATAC</a></code> object</p></dd>


<dt>regionGr</dt>
<dd><p><code>GRanges</code> object specifying the regions to aggregate over</p></dd>


<dt>samples</dt>
<dd><p>sample identifiers</p></dd>


<dt>countAggrFun</dt>
<dd><p>aggration function to be used for summarizing the insertion counts at each position. Possible values include <code>"sum"</code>, <code>"mean"</code>, and <code>"median"</code></p></dd>


<dt>norm</dt>
<dd><p>method used for normalizing the resulting position-wise counts.
Currently only <code>'tailMean'</code> is supported, which computes normalization factors as the mean signal in the tails of the window</p></dd>


<dt>normTailW</dt>
<dd><p>fraction of the region window to be used on each side of the window to be used for normalization if <code>norm</code> is one of <code>'tailMean'</code></p></dd>


<dt>kmerBiasAdj</dt>
<dd><p>compute Tn5 bias and use it to adjust the counts as in Corces, et al., Science, (2018)</p></dd>


<dt>k</dt>
<dd><p>length of the kmer to be used for sequence bias correction. Only relevant if <code>kmerBiasAdj==TRUE</code>.</p></dd>


<dt>sampleCovg</dt>
<dd><p>to save compute time, a sample coverage track list (as computed by <code>getCoverage(.object)</code>) can be supplied. If not, it will be computed on the fly.</p></dd>


<dt>sampleKmerFreqM</dt>
<dd><p>to save compute time, a matrix of sample kmer frequency at insertion sites (as computed by <code>getInsertionKmerFreq(.object, ...)</code>) can be supplied.
If not, it will be computed on the fly. Only relevant if <code>kmerBiasAdj==TRUE</code>.</p></dd>


<dt>regionKmerFreqM</dt>
<dd><p>to save compute time, a matrix of region kmer frequencies (kmers X window width).
Must have the same number of rows as the specified (or computed) <code>sampleKmerFreqM</code> (kmers)
and the same number of columns as the window width (median width of <code>regionGr</code>).
Only relevant if <code>kmerBiasAdj==TRUE</code>.</p></dd>


<dt>silent</dt>
<dd><p>limit log messages</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a <code>data.frame</code> containing position-wise counts (raw, normalized and optionally Tn5-bias-corrected) for each sample</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Fabian Mueller</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="http://epigenome-informatics.org" class="external-link">Fabian Mueller</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

