<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>setConfigElement — setConfigElement • ChrAccR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="setConfigElement — setConfigElement" />
<meta property="og:description" content="Set a configuration item to a given value" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ChrAccR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/overview.html">Overview</a>
    </li>
    <li>
      <a href="../articles/singlecell.html">Single-cell analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/articles/resource.html">Resource</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/GreenleafLab/ChrAccR">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>setConfigElement</h1>
    
    <div class="hidden name"><code>setConfigElement.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Set a configuration item to a given value</p>
    </div>

    <pre class="usage"><span class='fu'>setConfigElement</span><span class='op'>(</span><span class='va'>name</span>, <span class='va'>value</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>name</th>
      <td><p>name of the config item</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>value of the config item</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>nothing of particular interest.</p>
    <h2 class="hasAnchor" id="options-used-by-the-package"><a class="anchor" href="#options-used-by-the-package"></a>Options used by the package</h2>

    

<dl>
  <dt><b><code>tmpDir</code></b><code> = temdir()</code></dt><dd><p>Directory for temporary files. Must be existing.</p></dd>
  <dt><b><code>cleanMem</code></b><code> = TRUE</code></dt><dd><p>During runtime, regularly clean-out the memory in order to reduce memory overuse</p></dd>
  <dt><b><code>colorSchemes</code></b></dt><dd><p>named <code>list</code> of DISCRETE color schemes to be used for plotting. Each element should be a named vector specifying colors for groups/annotations.</p></dd>
  <dt><b><code>colorSchemesCont</code></b></dt><dd><p>named <code>list</code> of CONTINOUS color schemes to be used for plotting. Each element should be a vector specifying a range of colors.</p></dd>
  <dt><b><code>geneModelVersions</code></b></dt><dd><p>Gene model versions to be used for various genomes</p></dd>
  <dt><b><code>analysisName</code></b><code> = "ChrAccR analysis"</code></dt><dd><p>A title for the analysis (a string).</p></dd>
  <dt><b><code>regionTypes</code></b></dt><dd><p>Region types to be used in the analysis</p></dd>
  <dt><b><code>chromVarRegionTypes</code></b><code> = NULL</code></dt><dd><p>Region types to be used for chromVar analysis. If <code>NULL</code> (default), ChrAccR will automatically look for region types with the keyword <code>"peak"</code> in their name.</p></dd>
	 <dt><b><code>chromVarMotifs</code></b><code> = "jaspar_vert"</code></dt><dd><p>Character vector of names of TF motif sets to be used in ChromVAR analyses. By default the vertebrate set of the JASPAR database will be used.</p></dd>
  <dt><b><code>chromVarMotifNamesForDimRed</code></b></dt><dd><p>Names of motifs to be used for dimension reduction plots in the reports. [only relevant for single-cell data]</p></dd>
  <dt><b><code>genesOfInterest</code></b></dt><dd><p>Names of genes of interest to be highlighted in the reports (e.g. dimension reduction) in the reports.
      [currently only relevant for single-cell data and only when <code>scGeneActivity</code> is activated]</p></dd>
  <dt><b><code>annotationColumns</code></b></dt><dd><p>Sample annotation columns to be used for reporting</p></dd>
  <dt><b><code>annotationMinGroupSize</code></b><code> = 2</code></dt><dd><p>Minimum size of a group to be used in the reports. Influences which annotation columns are automatically selected for reporting.</p></dd>
  <dt><b><code>annotationMaxGroupCount</code></b><code> = NULL</code></dt><dd><p>Maximum number of groups to be used in the reports. Influences which annotation columns are
      automatically selected for reporting. If <code>NULL</code> (default) it will effectively be the
      number of samples - 1.</p></dd>
  <dt><b><code>doPeakCalling</code></b><code> = FALSE</code></dt><dd><p>Perform per-sample peak calling and retrieve consensus peak set. Requires that <code>macs2</code> is installed and can be called from the command line. [for bulk data analysis only]</p></dd>
  <dt><b><code>annotationPeakGroupColumn</code></b></dt><dd><p>Annotation column to base the consensus peak set replication filtering on.</p></dd>
  <dt><b><code>annotationPeakGroupAgreePerc</code></b><code> = 1.0</code></dt><dd><p>Percent of samples that have to agree to identify consensus peaks. See <code><a href='getConsensusPeakSet.html'>getConsensusPeakSet</a></code> for details.</p></dd>
  <dt><b><code>filteringCovgCount</code></b><code> = 1L</code></dt><dd><p>Minimum insertion count to filter count matrices by. See <code><a href='filterLowCovg,DsATAC-method.html'>filterLowCovg,DsATAC-method</a></code> for details. [for bulk data analysis only]</p></dd>
  <dt><b><code>filteringCovgReqSamples</code></b><code> = 0.75</code></dt><dd><p>Minimum required samples to apply low coverage filtering to. See <code><a href='filterLowCovg,DsATAC-method.html'>filterLowCovg,DsATAC-method</a></code> for details. [for bulk data analysis only]</p></dd>
  <dt><b><code>filteringSexChroms</code></b><code> = FALSE</code></dt><dd><p>Flag indicating whether to remove sex chromosomes.</p></dd>
  <dt><b><code>filteringScMinFragmentsPerCell</code></b><code> = 1000L</code></dt><dd><p>Minimum number of fragments per cell to retain a cell in the analysis. [for single-cell data analysis only]</p></dd>
  <dt><b><code>filteringScMaxFragmentsPerCell</code></b><code> = Inf</code></dt><dd><p>Maximum number of fragments allowed per cell to retain a cell in the analysis. [for single-cell data analysis only]</p></dd>
  <dt><b><code>filteringScMinTssEnrichment</code></b><code> = 6</code></dt><dd><p>Minimum TSS enrichment score per cell to retain a cell in the analysis. [for single-cell data analysis only]</p></dd>
  <dt><b><code>normalizationMethod</code></b><code> = "quantile"</code></dt><dd><p>Normalization method to use for count normalization. Allowed methods include the ones listed in <code><a href='transformCounts,DsATAC-method.html'>transformCounts,DsATAC-method</a></code>. [for bulk data analysis only]</p></dd>
  <dt><b><code>exploratoryLogNormCounts</code></b><code> = TRUE</code></dt><dd><p>Should a log-normalization be applied in the exploratory plot sections of the reports (dimension reduction, heatmaps)</p></dd>
  <dt><b><code>exploratoryNSubsample</code></b><code> = 2e6</code></dt><dd><p>Number of regions to subsample in exploratory analysis in order to increase computational performance.</p></dd>
  <dt><b><code>differentialColumns</code></b></dt><dd><p>Sample annotation columns to be used for differential testing and reporting</p></dd>
  <dt><b><code>differentialColumns1vsAll</code></b></dt><dd><p>Sample annotation columns to be used for differential testing and reporting in a 1-vs-all group setting. Should be a subset of <code>differentialColumns</code>.</p></dd>
  <dt><b><code>differentialCompNames</code></b></dt><dd><p>Comparison names from which comparison information is derived. Must be in the format of "$GRP1_NAME vs $GRP2_NAME [$ANNOTATION_COLUMN]".</p></dd>
  <dt><b><code>differentialAdjColumns</code></b></dt><dd><p>Sample annotation columns to be adjusted for in differential testing</p></dd>
  <dt><b><code>lolaDbPaths</code></b></dt><dd><p>Precomputed LOLA databases to be used for enrichment analysis. If <code>NULL</code> (default), ChrAccR will download an apropriate core database.</p></dd>
  <dt><b><code>scIterativeLsiRegType</code></b></dt><dd><p>For single-cell analysis only: region type to be used for clustering and dimension reduction using iterative LSI. By default (<code>NULL</code>),
      ChrAccR will look for a region type named <code>"tiling"</code>.</p></dd>
  <dt><b><code>scIterativeLsiParams</code></b></dt><dd><p>Parameters to use for iterative LSI. See <code><a href='iterativeLSI,DsATACsc-method.html'>iterativeLSI,DsATACsc-method</a></code> for details.</p></dd>
  <dt><b><code>scGeneActivity</code></b><code> = FALSE</code></dt><dd><p>For single-cell analysis only: Compute gene activity from accessibility.
      Possible options are <code>"RBF"</code> for radial-basis-function-weighted count aggregation (default when set to <code>TRUE</code>) or 
      <code>"Cicero"</code> for Cicero correlation-based aggregation</p></dd>

</dl>

    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Fabian Mueller</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://epigenome-informatics.org'>Fabian Mueller</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


